Add initial modifications to support the client credential grant

andrewbonney opened the pull request:
Work in progress to resolve #76

andrewbonney committed:
Add initial modifications to support the client credential grant

garethsb commented on APIs/AuthorizationAPI.raml:
```diff
@@ -173,7 +173,7 @@ types:
   post:
     description: |
-      Method for requesting OAuth 2.0 Bearer Token (Authorization Code and Refresh Grants).
+      Method for requesting OAuth 2.0 Bearer Token (Authorization Code, Client Credentials and Refresh Grants.)
```
*(Pedant alert)* Why's that full stop back inside the parentheses? :)

andrewbonney committed:
Add approval requirements for client credentials grant use

andrewbonney committed:
Add a requirement for restricting client credential use to scopes

andrewbonney commented:
We need to check what auth server implementations do when a client registration is present but disabled, as viewed from the token endpoint. We need to make sure that in this state a client implements backoff until they have been enabled, rather than continually attempting to re-register new client identifiers with the auth server.

cristian-recoseanu commented:
We also need to see if "**Disabling dynamic registered clients by default**" is a common functionality in off the shelf authorisation servers (achieved through a policy or other means). Basically we need to make sure our specified manual check for dynamic clients can take place in most products you can buy.

prince-chrismc commented on docs/4.1. Behaviour - Authorization Servers.md:
```diff
@@ -36,14 +36,20 @@ client type and redirect URIs to be provided.
+Authorization Servers MUST support authorization of dynamic client registrations via manual operator approval and MAY
+support additional mechanisms including:
+*   Presentation of a valid Client TLS certificate
+*   An initial JSON Web Token (JWT) as suggested in Section 3 of [RFC 7591][RFC-7591]
```
> Alternatively, if the server supports authorized registration, the
   developer or the client will be provisioned with an initial access
   token.  (The method by which the initial access token is obtained is
   out of scope for this specification.)

Have anyone ever tested this? How can a node tell if this is required?

garethsb replied:
A user generates an initial access token and installs it on the Node via some vendor-specific mechanism. The Node presents it to the Auth Server. So both Auth Server and Node would have to support this, which I don't think we should mandate. It's just another option. Keycloak seems to support it - https://www.keycloak.org/docs/latest/securing_apps/#_initial_access_token. Something to try out!

prince-chrismc replied:
There's another RFC very much related to this, I mentioned [here](https://github.com/AMWA-TV/nmos-authorization/issues/62#issuecomment-693609891)

andrewbonney replied:
The use of an initial access token is interesting, but perhaps not in the way Keycloak requires it at present. The dynamic client registration RFC talks about a 'software statement' JWT which I had somewhat dismissed previously due to bootstrapping concerns, but having looked again this has parallels with the client certificate mechanism we're using in the EST spec. A token built into the software and signed by the vendor could be used as the initial authorization mechanism, where the Authorization Server simply verifies the signature.

peterbrightwell replied:
Discussed on call.  Further checks on Keycloak mechanism and how well it's supported needed or alternatively more investigation of initial access token.

peterbrightwell replied:
Discussed on call. Support for software statement on OpenID's list of certifiers not currently good (though info is old) -- could be preferred method with fallback methods allowed.   To be continued...

andrewbonney committed:
Add requirements for client registration authentication methods

andrewbonney committed:
Add requirements around inclusion of client names during registration

andrewbonney commented:
I've added some suggested text based on what we've discussed around an order of preference for authentication during client registration. This also attempts to tackle #78 

One open question I have is whether Authorization Servers can typically be configured to have different authentication requirements dependent upon the `grant_types` requested. If they can then the requirements placed upon client using the `authorization_code` grant can be minimised, but if they typically can't then all clients will need to support the full set of authentication methods to maximise interoperability.

prince-chrismc commented on docs/4.2. Behaviour - Clients.md:
```diff
@@ -20,13 +20,26 @@ containing the required details).
+*   Unauthenticated registration
+
+Clients MUST be configurable to choose between the above authentication methods, which are listed in order of
+preference. Public keys which can be used to verify software statement signatures MUST be made available for end users
```
Is this plug and play enough?

andrewbonney replied:
The intent was to make sure that the software statement is the default as this is the 'most plug and play' method. That's only indicated by it being top of the order of preference though, so if it would benefit from being more explicit I can adjust.

prince-chrismc replied:
The main concern was installing public keys for each device in the auth server, but this is equivalent to the EST setup, so it's each vendor so it's only the first device that may need human configuration

andrewbonney commented:
A further question that comes to mind is how an Authorization Server can verify that a software statement belongs to the client which is providing it. As a software statement is likely to be signed at manufacture time it seems feasible that a statement from a trusted vendor could be intercepted and re-used by a 'rogue' device in order to gain access. This is made harder for pure software systems as the statement cannot be generated uniquely at manufacture for each instance which may be distributed.

peterbrightwell commented:
Update on today's call: more evidence of support for software statements has been found in new version of connect2id, including possible use in banking sector, but it's not yet clear how much.

Related but separate issue for machine-to-machine authentication: Possible extra security mechanism for client credential grant proposed on call: OpenID connect "Client assertion" authentication method -- client has to use private_key_jwt to connect to token endpoint.  This appears to be a preferred method for banking sector ("Open Banking" specification -- we don't know the level of support, and it's for banks internal use?). This would not necessarily help answer the bulk auth case (whereas software statements would).

Most servers seem to support private_key_jwt... can be seen as a form of software statement used for a similar purpose

Simon will be doing some tests over the next week

Noted: we are using most aspects of OAuth 2.0 that are in OpenID but not all are relevant.

Possible deficiency with software statement approach: could copy signed token between devices?

andrewbonney committed:
Remove software statement requirements from v1.0

These may be introduced in the future, but due to concerns around
their implementation in commercial auth servers and challenges in
testing it seems best to leave them out for now. There is also a
potential concern around the ability to steal and re-use software
statements.

andrewbonney committed:
Add requirements for clients and servers for token authentication

andrewbonney commented:
I've pushed two new commits.

- The first removes software statements for now. Whilst these seem like the best option for 'zero configuration' startup, the lack of implementations to test with presents a problem, and I have a concern around how straightforward it would be to steal and re-use a statement.
- The second is a first pass at indicating a preference to use the 'private_key_jwt' authentication method when interacting with the token endpoints and using confidential clients (those which have a client secret).

dannymeloy commented on docs/4.2. Behaviour - Clients.md:
```diff
@@ -20,22 +20,42 @@ containing the required details).
-Clients operating with a client secret SHOULD support using HTTP Basic Authentication, as per Section 2 of
-[RFC 2617][RFC-2617], to authenticate with the Authorization Server in the manner described in Section 2.3.1 of
-[RFC 6749][RFC-6749].
+Confidential Clients SHOULD support authentication using a digitally signed JSON Web Token, to authenticate with the
```
It might be best to highlight to vendors the need for clients to be able to support hosting their own public keys for validation of JWTs used for authentication. Something along the lines of:
```Clients SHOULD be capable of hosting their own public JWKs at a location that corresponds to the value of the registered jwks_uri URL used when registering with the Auth Server, if it was supplied```

lo-simon replied:
Also should memtion `jwks_uri` is preferred over the `jwks`, allowing easier key rotation.

andrewbonney committed:
Update as-per review comments around private_key_jwt

prince-chrismc commented on docs/4.2. Behaviour - Clients.md:
```diff
@@ -47,7 +47,10 @@ presence of the keys `registration_client_uri` and `registration_access_token` i
 This authentication method is identified as `private_key_jwt` in the `token_endpoint_auth_method` string during client
-registration.
+registration. Clients SHOULD be capable of hosting their own public JWKs at a location that corresponds to the value of
+the registered `jwks_uri` used when registering with the Authorization Server. Clients are RECOMMENDED to use the
+`jwks_uri` in preference to `jwks` such that their public keys can be replaced without re-registering with the
+Authorization Server.
```
If a node re-registered, if would presumably get a second client ID? If this is the case then a node would no longer be able to update the registry 🤔 Perhaps this may need to be a MUST?

andrewbonney replied:
In my head I was considering if a client may choose to use https://tools.ietf.org/html/rfc7592 if the server implemented it in this case, but as it's only a MAY for the auth server to implement that you may be right that we're better going that way.

prince-chrismc replied:
https://tools.ietf.org/html/rfc7592#section-2.2 would be helpful. It appears `jwks_uri` is more common too perhaps for a reason related to this.

andrewbonney replied:
Hopefully that's a little better now.

lo-simon reviewed (approved):
LGTM

andrewbonney committed:
Address review comment around preference for jwks_uri

garethsb reviewed (approved):
:clap:

peterbrightwell commented:
Approved on workshop review call
